### PR TITLE
feat: CSS selector for iOS

### PR DESCRIPTION
**Pull request by dpgraham**
Copies the same logic as what was done in UiAutomator2 Driver but makes some changes to make it work with iOS Class Chain.

### CODE REVIEW

File: lib/css-converter.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+  const val = css.value?.toLowerCase() || 'true'; // an omitted boolean attribute means 'true' (e.g.: input[checked] means checked is true)
+  switch (val) {
+    case '0':
+    case 'false':
+      return '0';
+    case '1':
+    case 'true':
+      return '1';
+    default:
+      throw new Error(`'${css.name}' must be true, false or empty. Found '${css.value}'`);
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:73]: TypeError

File: lib/css-converter.js [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+/**
+ * Convert a CSS attribute into a UiSelector method call
+ *
+ * @param {CssAttr} cssAttr CSS attribute object
+ * @returns {string} CSS attribute parsed as UiSelector
+ */
+function parseAttr (cssAttr) {
+  if (cssAttr.valueType && cssAttr.valueType !== 'string') {
+    throw new Error(`'${cssAttr.name}=${cssAttr.value}' is an invalid attribute. ` +
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:119]: TypeError

File: package.json [hunk 3]
```diff
@@ -55,12 +55,13 @@
     "appium-xcode": "^3.8.0",
     "async-lock": "^1.0.0",
     "asyncbox": "^2.3.1",
-    "bluebird": "^3.1.1",
     "js2xmlparser2": "^0.2.0",
     "lodash": "^4.17.10",
     "moment": "^2.24.0",
     "moment-timezone": "0.5.32",
     "node-simctl": "^6.4.0",
+    "bluebird": "^3.5.1",
```
Comment (mykola-mokhnach) [inline package.json:63]: imports should be sorted alphabetically

File: lib/css-converter.js [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+ * Convert a CSS selector to a iOS Class Chain selector
+ * @param {string} cssSelector CSS Selector
+ * @returns {string} The CSS selector converted to an iOS Class Chain
+ */
+CssConverter.toIosClassChainSelector = function toIosClassChainSelector (cssSelector) {
+  let cssObj;
+  try {
+    cssObj = parser.parse(cssSelector);
+  } catch (e) {
+    throw new errors.InvalidSelectorError(`Invalid CSS selector '${cssSelector}'. Reason: '${e}'`);
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:294]: e.message

File: lib/css-converter.js [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    case 'rule':
+      return parseCssRule(css);
+    case 'ruleSet':
+      return parseCssObject(css.rule);
+    case 'selectors':
+      return css.selectors.map((selector) => parseCssObject(selector)).join('; ');
+
+    default:
+      // This is never reachable, but if it ever is do this.
+      throw new Error(`iOS Class Chain does not support '${css.type}' css. Only supports 'rule', 'ruleSet', 'selectors' `);
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:280]: extra trailing space

File: lib/css-converter.js [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+/**
+ * Get the canonical form of a CSS attribute name
+ *
+ * Converts to lowercase and if an attribute name is an alias for something else, return
+ * what it is an alias for
+ *
+ * @param {Object} css CSS object
+ * @returns {string} The canonical attribute name
+ */
+function assertGetAttrName (css) {
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:86]: how about `requireAttributeName`?

File: lib/css-converter.js [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+ * @property {?string} value The value of the CSS object
+ */
+
+/**
+ * Get the boolean from a CSS object. If empty, return true. If not true/false/empty, throw exception
+ *
+ * @param {CssNameValueObject} css A CSS object that has 'name' and 'value'
+ * @returns {string} Either 'true' or 'false'. If value is empty, return 'true'
+ */
+function assertGetBool (css) {
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:63]: how about requireBoolean?

File: lib/css-converter.js [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+  ['index', ['nth-child']],
+];
+
+/**
+ * Convert hyphen separated word to snake case
+ *
+ * @param {string} str
+ * @returns {string} The hyphen separated word translated to snake case
+ */
+function toSnakeCase (str) {
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:42]: I thought snake-case is a string separated with underscores `_`. Hyphen-separated strings don't have a special name

File: lib/css-converter.js [hunk 9]
```diff
@@ -0,0 +1,303 @@
+import { CssSelectorParser } from 'css-selector-parser';
+import { escapeRegExp, find, filter, isString, isObject } from 'lodash';
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:2]: we usually import lodash as `_`

File: lib/css-converter.js [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+    for (const attr of cssRule.attrs) {
+      attrs.push(parseAttr(attr));
+    }
+  }
+  if (cssRule.pseudos) {
+    for (const pseudo of cssRule.pseudos) {
+      attrs.push(parsePseudo(pseudo));
+    }
+  }
+  const nonIndexAttrs = filter(attrs, (attr) => isString(attr));
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:239]: why do we need lodash's filter and find if javascript arrays have their own?

**Comment by mykola-mokhnach** [comment]
It would make sense to add this locator to README

File: lib/css-converter.js (hunk 8 above)
Comment (dpgraham) [reply lib/css-converter.js]: I changed it to camel case.

**Comment by dpgraham** [comment]
@mykola-mokhnach the README.md doesn't talk about selectors. I'll add this to the Appium site docs.

File: test/unit/css-converter-specs.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
+import chai from 'chai';
```
Comment (KazuCocoa) [inline test/unit/css-converter-specs.js:1]: 👍

File: lib/css-converter.js [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+ */
+
+/**
+ * Convert a CSS pseudo class to a UiSelector
+ *
+ * @param {CssPseudo} cssPseudo CSS Pseudo class
+ * @returns {string} Pseudo selector parsed as UiSelector
+ */
+function parsePseudo (cssPseudo) {
+  if (cssPseudo.valueType && cssPseudo.valueType !== 'string') {
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:175]: `!_.isString(cssPseudo.valueType)` could be used

File: lib/css-converter.js [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+    throw new Error(`'${cssPseudo.name}=${cssPseudo.value}'. ` +
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:176]: TypeError

File: lib/css-converter.js [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+  const indexAttr = attrs.find((attr) => _.isObject(attr) && attr.index);
+  if (indexAttr) {
+    iosClassChainSelector += `[${indexAttr.index}]`;
+  }
+
+  if (cssRule.rule) {
+    iosClassChainSelector += `/${parseCssRule(cssRule.rule)}`;
+  }
+
+  if (cssRule.nestingOperator === '>') {
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:253]: how about using a ternary operator?

File: lib/css-converter.js [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+    return iosClassChainSelector;
+  } else {
+    return `**/` + iosClassChainSelector;
```
Comment (mykola-mokhnach) [inline lib/css-converter.js:256]:
```
`**/${iosClassChainSelector}`
```

**Review by mykola-mokhnach — APPROVED** [review]

**Pull request merged**